### PR TITLE
Invoices service

### LINKED ISSUE #267: Add methods related to Invoice API
https://developers.scalingo.com/invoices

### PR DESCRIPTION
**Pull request by SCedricThomas**
fix #267

### CODE REVIEW

File: invoices.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+)
+
+type InvoicesService interface {
+	ListInvoices(context.Context, PaginationOpts) (Invoices, PaginationMeta, error)
+	InvoiceShow(context.Context, string) (*Invoice, error)
+}
+
+var _ InvoicesService = (*Client)(nil)
+
+const LayoutISO = "2006-01-02"
```
Comment (Soulou) [inline invoices.go:18]: This constant name is not right

```suggestion
const BillingMonthDateFormat = "2006-01-02"
```

File: invoices.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+type invoiceDate time.Time
```
Comment (Soulou) [inline invoices.go:20]: Here too, to be consistant with the rest

```suggestion
type billingMonthDate time.Time
```

**Review by Soulou — CHANGES_REQUESTED** [review]
My suggestion plus according changes in the rest of the code

File: invoices.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+func (c *Client) InvoiceShow(ctx context.Context, id string) (*Invoice, error) {
+	var invoiceRes InvoiceRes
+	err := c.ScalingoAPI().ResourceGet(ctx, "account/invoices", id, nil, &invoiceRes)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	return invoiceRes.Invoice, nil
+}
+
+func (i *billingMonthDate) UnmarshalJSON(b []byte) error {
```
Comment (Soulou) [inline invoices.go:79]: Sorry to be picky, why `i` ? :D

File: invoices.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	value := strings.Trim(string(b), `"`)
+	if value == "" || value == "null" {
+		return nil
+	}
+	var t time.Time
```
Comment (Soulou) [inline invoices.go:84]: We don't use this syntax but

```
t, err := time.Parse(...)
if err != nil {
 ...
}
```

File: invoices.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	var err error
+	if t, err = time.Parse(BillingMonthDateFormat, value); err != nil {
+		return err
+	}
+	*i = billingMonthDate(t)
+	return nil
+}
+
+func (i billingMonthDate) MarshalJSON() ([]byte, error) {
```
Comment (Soulou) [inline invoices.go:93]: We actually never need to marshal into JSON, just read from the API, so this method is useless, I would delete it. (it's dead code)

**Review by Soulou — CHANGES_REQUESTED** [review]
Sorry picky to teach good practices

File: invoices.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import (
+	"context"
+	"strings"
+	"time"
+
+	"gopkg.in/errgo.v1"
+)
+
+type InvoicesService interface {
+	ListInvoices(context.Context, PaginationOpts) (Invoices, PaginationMeta, error)
```
Comment (EtienneM) [inline invoices.go:12]: From what I remember this is how we name the list functions in go-scalingo.

nitpick:
```suggestion
	InvoicesList(context.Context, PaginationOpts) (Invoices, PaginationMeta, error)
```

File: invoices.go (hunk 3 above)
Comment (SCedricThomas) [reply invoices.go:79]: Because I was out of phase yesterday afternoon ! x)

**Review by Soulou — APPROVED** [review]

**Pull request merged**